csharp/src/Google.Protobuf/Collections/RepeatedField.cs

```diff
@@ -101,10 +101,11 @@ public void AddEntriesFrom(CodedInputStream input, FieldCodec<T> codec)
                 int length = input.ReadLength();
                 if (length > 0)
                 {
+                    EnsureSize(count + length);
```

Comment:
While I was thinking about adding "streaming support" I found a nice solution for this reallocation issue.
It requires modifying codegen for C# by enabling use of partial method (as i intended for streaming support).
```
        public void MergeFrom(pb::CodedInputStream input)
        {
            uint tag;
            while ((tag = input.ReadTag()) != 0)
            {
                switch (tag)
                {
                    default:
                        input.SkipLastField();
                        break;
[...]
                    case 18:
                        {
//  We can insert partial method call here -------->
                            objects_.AddEntriesFrom(input, _repeated_objects_codec);
                            break;
                        }
                }
            }
        }
```

I intended to use this partial method to handle objects decoded from stream. Deserialize straight into destination structures instead of temporal protobuf representation.

But I have just realized that when developer wants to do a small size-to-performance tradeoff. He/she can insert an additional int32 field for an array length hint (a bit redundant) but it would allow to preallocate proper size of buffer here. Of course for security reasons it should not be treated authoritatively (or have a builtin support!) however it can be a nice preallocation hint.

This field would be:
- optional
- will work with both packed and unpacked formats
- will not require any special support from proto3, even not declaring custom options
- could be utilized on per repeated INSTANCE (not even type) basis (e.g. only if there is a significant number of repetitions)